<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Herring Selective Sweep</title>
    <meta charset="utf-8" />
    <meta name="author" content="Love Chrisson, Felix Lenner &amp; Ben Titmus" />
    <meta name="date" content="2021-05-27" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Herring Selective Sweep
## Bioinformatics 3MR103
### Love Chrisson, Felix Lenner &amp; Ben Titmus
### 2021-05-27

---





# Introduction/Background 

- whats a sweep

- the fishes maybe

- other stuff...

- Aim

- "A brief introduction to the project (What is it about? What is the purpose of the project?"

---

# Methods Overview

- "A methods overview (describe briefly, e.g using bullet points, what you have done and what data you have used)"

---

# Detailed methods and results: 

- "Walk us through important code snippets or command line calls that you have used"

- "Present what is the input and output, or intermediate results, of the code snippet or command line call you describe."

---

# Check population structure 

1. Extract chromosome 4 

```bash
vcftools --gzvcf data/Ch_v2.0.2_79_ind_snpEff.vcf.gz 
--chr chr4 --recode --out data/chr4
```

1. Principal component analysis with PLINK 

```bash
plink --vcf data/Ch_v2.0.2_79_ind_snpEff_chr4.vcf 
--double-id  --pca tabs --out data/chr4_pca
```
---
# Plot PCA
- Read and import the files

```r
library(tidyverse)
# Read files
eigenvec &lt;- read_tsv("data/chr4_pca.eigenvec", col_names = F)
eigenval &lt;- scan("data/chr4_pca.eigenval") 

# Set eigenvec column names
colnames(eigenvec) &lt;- c("ID1", "ID2", 
                        paste0("PC", 1:(ncol(eigenvec)-2)))

# List areas
areas &lt;- c("Atlantic", "Baltic", "SeaofJapan", "BarentsSea", "WhiteSea", "Pacific")

# extract strings from ID1, if it matches any of the ares, put corresponding value in area
eigenvec &lt;- eigenvec %&gt;% 
  mutate(Area = str_extract(ID1, paste(areas, collapse = "|")))
```
---
# Construct variance label

```r
variance_label &lt;- function(pc, var_vector) {
  # keep PC number
  pc &lt;- as.numeric(sub("\\D+", "", pc))
  # Calculate variance
  variance &lt;- var_vector[pc] / sum(var_vector)
  # Construct label
  label &lt;- paste0("PC", pc, " (", round(variance*100, 2), "%)")
  # Return the label 
  return(label)
}
```
---
# Plot PCA with area as color 

```r
ggplot(eigenvec, aes(x = PC1, y = PC2, color = Area)) + 
  geom_point() +
  xlab(variance_label("PC1", eigenval)) + 
  ylab(variance_label("PC2", eigenval))
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-4-1.svg)&lt;!-- --&gt;
---
# Balsfjord are outliers


```r
# Balsfjord are outliers in Atlantic 
eigenvec %&gt;% filter(PC1 &gt; 0 &amp; Area == "Atlantic") %&gt;% select(ID1)
```

```
## # A tibble: 4 x 1
##   ID1                     
##   &lt;chr&gt;                   
## 1 HWS61_Balsfjord_Atlantic
## 2 HWS62_Balsfjord_Atlantic
## 3 HWS63_Balsfjord_Atlantic
## 4 HWS64_Balsfjord_Atlantic
```
---
# Extract samples, produce new vcf files 

```bash
# Get samples using bcftools
bcftools query -l chr4.recode.vcf | grep "Baltic" &gt; baltic_samples.txt
bcftools query -l .recode.vcf | grep "Atlantic" | grep -v "Balsfjord" &gt; atlantic_samples.txt

# Produce vcf-files for chr4 where atlantic, baltic, atlantic and balic are kept
vcftools --vcf chr4.recode.vcf -keep atlantic_samples.txt --recode --out chr4_atlantic
vcftools --vcf chr4.recode.vcf -keep baltic_samples.txt --recode --out chr4_baltic
# VCFtools can be used to extract allele frequencies
vcftools --vcf data/chr4_atlantic.recode.vcf --freq2 --out chr4_atlantic
vcftools --vcf data/chr4_baltic.recode.vcf --freq2 --out chr4_baltic
```
---
# Load allele frequencies into R

```r
# Set column names and types 
allele_column_names = c("CHROM", "POS", "N_ALLELES", "N_CHR", "A1_FREQ", "A2_FREQ")
allele_column_types = "cddddd"

# Load samples
atlantic_raw_allele_freq &lt;- read_tsv("data/chr4_atlantic.frq", skip = 1, col_names = allele_column_names, col_types = allele_column_types)
baltic_raw_allele_freq &lt;- read_tsv("data/chr4_baltic.frq", skip = 1, col_names = allele_column_names, col_types = allele_column_types)

# Select important, add identifier
atlantic_allele_freq &lt;- atlantic_raw_allele_freq %&gt;% 
  select(POS, FREQ = A2_FREQ) %&gt;% 
  mutate(ID = "Atlantic Alt Allele Freq")

baltic_allele_freq &lt;- baltic_raw_allele_freq %&gt;% 
  select(POS, FREQ = A2_FREQ) %&gt;% 
  mutate(ID = "Baltic Alt Allele Freq")

# Make delta df, calculate delta af
delta_allele_freq &lt;- atlantic_allele_freq %&gt;% select(POS)

delta_allele_freq &lt;- delta_allele_freq %&gt;% 
  mutate(FREQ = abs(atlantic_allele_freq$FREQ - baltic_allele_freq$FREQ), 
         ID = "Delta Allele Freq")
```
---
# More R

```r
combined_allele_freq &lt;- rbind(atlantic_allele_freq, 
                       baltic_allele_freq, 
                       delta_allele_freq)

window_size = 1000

# Calculates break size based on window size 
break_size &lt;- function(positions, window_size) {
  breaks &lt;- (max(positions) - min(positions)) / (window_size)
  return(breaks)
}
```
---
# Plot alleles in R

```r
combined_allele_freq %&gt;%
  group_by(gr = cut(POS, break_size(POS, window_size)), ID) %&gt;%
  summarize(`Position (Mb)` = mean(POS/1e6), `Mean Frequency` = mean(FREQ)) %&gt;%
  ggplot(aes(`Position (Mb)`,`Mean Frequency`, col = ID)) + 
  geom_jitter(width = 0, height = 0.03, size = .1, alpha = .3) + 
  facet_wrap(~ID, ncol = 3)
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-6-1.svg)&lt;!-- --&gt;
---

# Heterozygosity - not run 

```r
# Load heterozygosity 
atlantic_raw_het_freq &lt;- read_tsv("data/chr4_atlantic.hwe")
baltic_raw_het_freq &lt;- read_tsv("data/chr4_baltic.hwe")

# Separate `OBS(HOM1/HET/HOM2)` column, calculate frequency and select important
atlantic_het_freq &lt;- atlantic_raw_het_freq %&gt;% 
  separate(`OBS(HOM1/HET/HOM2)`, c("OBS_HOM1", "OBS_HET", "OBS_HOM2"), convert = T) %&gt;%
  mutate(TOT_OBS = OBS_HOM1 + OBS_HET + OBS_HOM2, 
         FREQ = OBS_HET / TOT_OBS,
         ID = "Atlantic Obs Het Freq") %&gt;%
  select(POS, FREQ, ID, TOT_OBS)

baltic_het_freq &lt;- baltic_raw_het_freq %&gt;% 
  separate(`OBS(HOM1/HET/HOM2)`, c("OBS_HOM1", "OBS_HET", "OBS_HOM2"), convert = T) %&gt;%
  mutate(TOT_OBS = OBS_HOM1 + OBS_HET + OBS_HOM2, 
         FREQ = OBS_HET / TOT_OBS,
         ID = "Baltic Obs Het Freq") %&gt;%
  select(POS, FREQ, ID, TOT_OBS)

# Initialize delta df with POS
delta_het_freq &lt;- atlantic_het_freq %&gt;% select(POS)

# Calculate delta 
delta_het_freq &lt;- delta_het_freq %&gt;% 
  mutate(FREQ = abs(atlantic_het_freq$FREQ - baltic_het_freq$FREQ), 
         ID = "Delta Obs Het Freq")

atlantic_baltic_het_freq &lt;- rbind(atlantic_het_freq, baltic_het_freq)
```
---
# Missingness plot - this would be quite nice I guess

```r
# Should be in code above but needs to be run
# Choose a filter threshold, % of max
filter_threshold = .95

# Keep positions 
positions_to_keep &lt;- inner_join(atlantic_het_freq, baltic_het_freq, by = "POS") %&gt;%
  filter(TOT_OBS.x &gt;= filter_threshold * max(TOT_OBS.x) &amp;
         TOT_OBS.y &gt;= filter_threshold * max(TOT_OBS.y)) %&gt;%
  select(POS)

atlantic_baltic_het_freq %&gt;%
  group_by(gr = cut(POS, break_size(POS, 1e5)), ID) %&gt;%
  summarize(`Position (Mb)` = mean(POS/1e6), `Mean missingness` = mean(1-TOT_OBS/max(TOT_OBS))) %&gt;%
  ggplot(aes(`Position (Mb)`,`Mean missingness`, col = ID)) + 
  geom_line()
```

```
## `summarise()` has grouped output by 'gr'. You can override using the `.groups` argument.
```

![](herring_project_presentation_files/figure-html/missingness-plot-1.png)&lt;!-- --&gt;

```r
# Remove the TOT_OBS column to be able to combine with delta_het_freq and plot without changing the code
atlantic_het_freq2 &lt;- atlantic_het_freq %&gt;% select(POS, FREQ, ID)
baltic_het_freq2 &lt;- baltic_het_freq %&gt;% select(POS, FREQ, ID)
```
---
# Annotation and code stuff
- Bash commands to get annotations here maybe... 

```r
# *** ANNOTATION ***

annotate_positions &lt;- function(input_df, annotation_df) {
  # Takes an input dataframe containing (at least) POS, and annotation df containing POS and ANN, 
  # and appends the annotations to the input dataframe
  
  # Filter positions which should be annotated 
  filtered_annotations &lt;- annotation_df %&gt;% filter(POS %in% input_df$POS)
  
  # This info can be found in the VCF file 
  snpEff_column_names &lt;- c("Allele","Annotation","Annotation_Impact","Gene_Name","Gene_ID","Feature_Type","Feature_ID","Transcript_BioType","Rank","HGVS.c","HGVS.p","cDNA.pos / cDNA.length","CDS.pos / CDS.length","AA.pos / AA.length","Distance","ERRORS / WARNINGS / INFO")
  
  untangled_annotations &lt;- filtered_annotations %&gt;%
    mutate(ANN = str_split(ANN, ",")) %&gt;% # Split the ANN column into a list of annotations by comma
    unnest(ANN) %&gt;% # create rows for each position - annotation pair
    group_by(POS) %&gt;% # group by position 
    mutate(Annotation_ID = seq_along(POS), .after = POS) %&gt;% # assign a number to each annotation
    ungroup() %&gt;% # ungroup the values again 
    #separate is really slow in R 4.x for some reason, maybe some map variant would be better?
    separate(ANN, into = snpEff_column_names, sep = "\\|")# separate the ANN column (now with one annotation per row) into columns based on SnpEff information
  
  # join annotations to input df
  output_df &lt;- inner_join(input_df, untangled_annotations, by = "POS")  
  
  return(output_df)
}

library(vroom) # you could also use read_delim, but its (~9x) slower in this case

# bcftools query -f "%POS\t%ANN\n" chr4.vcf &gt; original_already_annotated.txt
annotation &lt;- vroom("data/original_already_annotated.txt", delim = "\t", col_names = c("POS", "ANN"))
```

```
## Rows: 1,158,327
## Columns: 2
## Delimiter: "\t"
## chr [1]: ANN
## dbl [1]: POS
## 
## Use `spec()` to retrieve the guessed column specification
## Pass a specification to the `col_types` argument to quiet this message
```

```r
# Filter on frequency, takes to long time to annotate all positions and we don't need to 
# Or a certain part of the chromosome, .9e7 to 1.4e7 works fine to annotate ~
significant_positions &lt;- delta_allele_freq %&gt;% filter(between(POS, 0.9e7, 1.3e7)) %&gt;% filter(FREQ &gt; 0.6)

# Remove SNPs with high missingess by only keeping those that are in (in my case) positions_to_keep
# (if you have not already done so),
# Before running the annotate_positions functions

annotated_positions &lt;- significant_positions %&gt;%
  filter(POS %in% positions_to_keep$POS) %&gt;%
  annotate_positions(annotation)

# Can then filter what we want, for example
annotated_positions %&gt;% group_by(Gene_ID) %&gt;%
  filter(Feature_Type == "transcript") %&gt;% 
  summarize(mean.POS = as.integer(mean(POS)), mean.FREQ = mean(FREQ)) %&gt;%
  select(POS = mean.POS, FREQ = mean.FREQ, Gene_ID)
```

```
## # A tibble: 7 x 3
##        POS  FREQ Gene_ID           
##      &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;             
## 1 11182246 0.633 ENSCHAG00000025774
## 2 11220004 0.632 ENSCHAG00000025873
## 3 11292244 0.652 ENSCHAG00000025879
## 4 11332637 0.643 ENSCHAG00000025886
## 5 11354827 0.638 ENSCHAG00000025895
## 6 11381259 0.645 ENSCHAG00000026024
## 7 11404399 0.639 ENSCHAG00000026767
```

```r
# A bug in biomaRt overwrites dplyr::select, workaround is to unload dplyr, then load it _after_ biomaRt
detach("package:dplyr")
library(biomaRt)
```

```
## Warning: package 'biomaRt' was built under R version 4.0.3
```

```r
library(dplyr)
```

```
## 
## Attaching package: 'dplyr'
```

```
## The following object is masked from 'package:biomaRt':
## 
##     select
```

```
## The following objects are masked from 'package:stats':
## 
##     filter, lag
```

```
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
```

```r
# get the herring database
ensembl = useMart("ensembl",dataset="charengus_gene_ensembl")

zfin_ids &lt;- annotated_positions %&gt;% 
  filter(Feature_Type == "transcript" &amp; Annotation != "intron_variant") %&gt;%
  group_by(Gene_ID) %&gt;%
  dplyr::select(POS, FREQ, Gene_ID, Annotation_ID, Feature_Type, Annotation) %&gt;%
  summarize(mean.pos = mean(POS), mean.freq = mean(FREQ), Gene_ID) %&gt;% unique() %&gt;%
  mutate(getBM(attributes=c('ensembl_gene_id', 'zfin_id_symbol'), # Get zfin_id_symbol
                    filters = 'ensembl_gene_id', 
                    values = Gene_ID, 
                    mart = ensembl, uniqueRows = F)) %&gt;%
  mutate(zfin_id_symbol = ifelse(zfin_id_symbol == "", NA, zfin_id_symbol)) %&gt;%
  #na.omit() %&gt;% # if we want to filter only those genes which has zfin_ids
  dplyr::select(-ensembl_gene_id)
```

```
## `summarise()` has grouped output by 'Gene_ID'. You can override using the `.groups` argument.
```
---
# Combine data 

```r
# Combine data - Commented out combining all of the data 
combined_data &lt;- rbind(atlantic_allele_freq, 
                       #atlantic_het_freq2, 
                       baltic_allele_freq, 
                       #baltic_het_freq2, 
                       delta_allele_freq)
                       #delta_het_freq)
# Combine only the heterozygosity
#combined_data &lt;- rbind(atlantic_het_freq2, baltic_het_freq2, delta_het_freq)

# Filter out data
combined_data &lt;- combined_data %&gt;% filter(POS %in% positions_to_keep$POS)
```
---
# Summarize the final results

- Maybe a plot of delta allele freq with annotated genes here? 

- Or a plot + list of genes

```r
library(ggrepel)

combined_data %&gt;%
  filter(ID == "Delta Allele Freq") %&gt;%
  filter(between(POS, .9e7, 1.4e7)) %&gt;%
  #group_by(gr = cut(POS, break_size(POS, window_size))) %&gt;%
  #summarize(`Position (Mb)` = mean(POS/1e6), `Mean Frequency` = mean(FREQ)) %&gt;%
  mutate(`Position (Mb)` = POS/1e6, `Frequency` = FREQ) %&gt;%
  #ggplot(aes(`Position (Mb)`, `Mean Frequency`)) + 
  ggplot(aes(`Position (Mb)`, `Frequency`)) + 
  geom_jitter(alpha = .3, size = .3, width = 0, height =.0) + 
  geom_point(data = zfin_ids, aes(x = mean.pos/1e6, y = mean.freq, color = "red"), alpha = 1, size = .3) +
  geom_text_repel(data = zfin_ids, 
                  aes(x = mean.pos/1e6, y = mean.freq, label = zfin_id_symbol), 
                  max.overlaps = Inf, 
                  min.segment.length = 0,
                  segment.size = 0.2,
                  box.padding = 1
                  #segment.color = "red",
                  #arrow = arrow(length = unit(0.07, "inches"), type = "open")
                  ) + theme(legend.position = "none") + 
  ylab("Delta Allele Freq") + 
  xlab("Chromosome 4 (Mb)")
```

```
## Warning: Removed 3 rows containing missing values (geom_text_repel).
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-9-1.png)&lt;!-- --&gt;

```r
zfin_ids
```

```
## # A tibble: 7 x 4
## # Groups:   Gene_ID [7]
##   Gene_ID             mean.pos mean.freq zfin_id_symbol
##   &lt;chr&gt;                  &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;         
## 1 ENSCHAG00000025774 11184501.     0.617 magi1a        
## 2 ENSCHAG00000025873 11220005.     0.632 rho           
## 3 ENSCHAG00000025879 11326441      0.643 &lt;NA&gt;          
## 4 ENSCHAG00000025886 11329217.     0.643 &lt;NA&gt;          
## 5 ENSCHAG00000025895 11352886.     0.636 &lt;NA&gt;          
## 6 ENSCHAG00000026024 11383908.     0.646 qars1         
## 7 ENSCHAG00000026767 11383125      0.671 espn
```

---

# Conclusion

- "Conclusions (What do the results tell us?)."

- "Also comment on the project and the methods you used." 

- "Could you have done it in a different way? What difficulties did you encounter? etc."

---

class: center, top
background-image: url("https://upload.wikimedia.org/wikipedia/commons/c/c5/Clupea_harengus_Gervais.flipped.jpg")
background-size: contain

# Thanks!
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
