<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Herring Selective Sweep</title>
    <meta charset="utf-8" />
    <meta name="author" content="Love Chrisson, Felix Lenner &amp; Ben Titmuss" />
    <meta name="date" content="2021-05-31" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Herring Selective Sweep
## Bioinformatics 3MR103
### Love Chrisson, Felix Lenner &amp; Ben Titmuss
### 2021-05-31

---

&lt;style type="text/css"&gt;
pre {
  white-space: pre-wrap;
}
.nowrap pre{
  white-space: nowrap;
}
.tiny .remark-code {
  font-size: 60% !important;
}
&lt;/style&gt;





# Background 

- The Baltic Sea was formed approximately 15,000 years ago, and following its creation the Atlantic herring migrated into the new habitat causing a bottleneck of their gene pool and forming a population that would become the subspecies Baltic herring. 

- This event caused a selective sweep to take place in the new population.

- Selective Sweep = The process of a beneficial mutation increasing in frequency and becoming fixed, resulting in less genetic variation in the genome around the mutation.

- Selective sweeps can be identified by changes in heterozygosity and increases in allele frequency in regions.

- By determining the presence of selective sweeps in the chromosome, it should be possible to discover the genes that had to mutate (and then became “fixed” in the population) in order to allow the herring to adapt to their new environment. 
---

# Aim

Locate the peak and range of a selective sweep of interest on chromosome 4 in the genome of the studied baltic herring samples, and find out which gene/genes that have played a key role in the development of the new herring subspecies. 

---

# Methods Overview

- Received a VCF file with genomic data from multiple species/locations of herring samples. Including Atlantic, Baltic, Sea of Japan, White Sea, Pacific herring. 

- Extract all data related to chromosome 4 from the original file in order to have a file of a more manageable size to work with. 

- Used PCA analysis to study the provided data to better understand population structure

- Through a filtering process remove unwanted populations and samples from the study

- Calculate the heterozygosity of alleles along the chromosome

- Filter individuals who have a missingness value of 1

- Calculate the allele frequency and delta allele frequency across the chromosome

- Determine the peak and range of the selective sweep of interest    

- Annotate SNPs of interest 

---

# Check population structure 

Extract chromosome 4


```bash
vcftools --gzvcf data/Ch_v2.0.2_79_ind_snpEff.vcf.gz 
--chr chr4 --recode --out data/chr4
```

Principal component analysis with PLINK


```bash
plink --vcf data/Ch_v2.0.2_79_ind_snpEff_chr4.vcf 
--double-id  --pca tabs --out data/chr4_pca
```

???

Then read output into R and plot 

---

# Check population structure

Read the files into R

.tiny[

```r
library(tidyverse)

# Read files
eigenvec &lt;- read_tsv(file = eigenvec_path, col_names = F)
eigenval &lt;- scan(file = eigenval_path) 

# Set eigenvec column names
colnames(eigenvec) &lt;- c("ID1", "ID2", 
                        paste0("PC", 1:(ncol(eigenvec)-2)))

# List areas
areas &lt;- c("Atlantic", "Baltic", "SeaofJapan", "BarentsSea", "WhiteSea", "Pacific")

# extract strings from ID1, if it matches any of the ares, put corresponding value in area
eigenvec &lt;- eigenvec %&gt;% 
* mutate(Area = str_extract(ID1, paste(areas, collapse = "|")))
```

Calculate variance and create label


```r
variance_label &lt;- function(pc, var_vector) {
  # keep PC number
  pc &lt;- as.numeric(sub("\\D+", "", pc))
  # Calculate variance
  variance &lt;- var_vector[pc] / sum(var_vector)
  # Construct label
  label &lt;- paste0("PC", pc, " (", round(variance*100, 2), "%)")
  # Return the label 
  return(label)
}
```
]

---

# PCA Plot


```r
ggplot(eigenvec, aes(x = PC1, y = PC2, color = Area)) +
  # minor amount of jitter to make the baltic samples visible
  geom_jitter(width = 0.01, height = 0.03, alpha = 0.5) +
  xlab(variance_label("PC1", eigenval)) + 
  ylab(variance_label("PC2", eigenval))
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-5-1.png)&lt;!-- --&gt;
---
# Balsfjord herring are not Atlantic


```r
# Filter the PCA data for the atlantic outliers 

eigenvec %&gt;% filter(PC1 &gt; 0 &amp; Area == "Atlantic") %&gt;% select(ID1)
```

```
# # A tibble: 4 x 1
#   ID1                     
#   &lt;chr&gt;                   
# 1 HWS61_Balsfjord_Atlantic
# 2 HWS62_Balsfjord_Atlantic
# 3 HWS63_Balsfjord_Atlantic
# 4 HWS64_Balsfjord_Atlantic
```
---
# Allele frequencies

Get a list of the samples


```bash
# same for baltic
bcftools query -l chr4.vcf | grep "Atlantic" | grep -v "Balsfjord" &gt; atlantic_samples.txt

```

Output allele frequencies


```bash
# same for baltic
vcftools --vcf chr4.vcf --keep atlantic_samples.txt --freq2 --out chr4_atlantic
```

???

Then read into R

---

# Calculate allele frequencies


```bash
# Get sample names
bcftools query -l chr4.recode.vcf | grep "Baltic" &gt; baltic_samples.txt
bcftools query -l chr4.recode.vcf | grep "Atlantic" | grep -v "Balsfjord" &gt; atlantic_samples.txt

# Produce vcf-files for chr4 where atlantic and baltic are kept 
vcftools --vcf chr4.recode.vcf -keep atlantic_samples.txt --recode --out chr4_atlantic
vcftools --vcf chr4.recode.vcf -keep baltic_samples.txt --recode --out chr4_baltic

# VCFtools can be used to extract allele frequencies
vcftools --vcf data/chr4_atlantic.recode.vcf --freq2 --out chr4_atlantic
vcftools --vcf data/chr4_baltic.recode.vcf --freq2 --out chr4_baltic
```

---

# Read Allele frequencies into R


```r
allele_column_names = c("CHROM", "POS", "N_ALLELES", "N_CHR", "A1_FREQ", "A2_FREQ")
allele_column_types = "cddddd"

#Load samples 
atlantic_raw_allele_freq &lt;- read_tsv(file = atlantic_freq_path, skip = 1, col_names = allele_column_names, col_types = allele_column_types)

baltic_raw_allele_freq &lt;- read_tsv(file = baltic_freq_path, skip = 1, col_names = allele_column_names, col_types = allele_column_types)
```


```r
#Select important, add identifier
atlantic_allele_freq &lt;- atlantic_raw_allele_freq %&gt;% 
  select(POS, FREQ = A2_FREQ) %&gt;% 
  mutate(ID = "Atlantic Alt Allele Freq")

baltic_allele_freq &lt;- baltic_raw_allele_freq %&gt;% 
  select(POS, FREQ = A2_FREQ) %&gt;% 
  mutate(ID = "Baltic Alt Allele Freq")
```

---

# Calculate delta and window-size


```r
delta_allele_freq &lt;- atlantic_allele_freq %&gt;% select(POS)

#Calculate delta allele frequency
delta_allele_freq &lt;- delta_allele_freq %&gt;% 
  mutate(
*   FREQ = abs(atlantic_allele_freq$FREQ -
*                baltic_allele_freq$FREQ),
    ID = "Delta Allele Freq")
```


```r
# Combine data
combined_allele_freq &lt;- rbind(atlantic_allele_freq, 
                       baltic_allele_freq, 
                       delta_allele_freq)
# Set window size
*window_size = 1000

# Calculates break size based on window size 
break_size &lt;- function(positions, window_size) {
  breaks &lt;- (max(positions) - min(positions)) / (window_size)
  return(breaks) }
```

---
# Plot allele frequencies 


```r
combined_allele_freq %&gt;%
* group_by(gr = cut(POS, break_size(POS, window_size)), ID) %&gt;%
  summarize(`Position (Mb)` = mean(POS/1e6), `Mean Frequency` = mean(FREQ)) %&gt;%
  ggplot(aes(`Position (Mb)`,`Mean Frequency`, col = ID)) + 
  geom_jitter(width = 0, height = 0.03, size = .1, alpha = .3, show.legend = FALSE) + 
  facet_wrap(~ID, ncol = 3)
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-9-1.png)&lt;!-- --&gt;

---

# Heterozygosity 

Use VCFtools to get heterozygosity frequencies


```bash
vctools --vcf chr4.vcf --keep atlantic_samples.txt --hardy --out ...
```

Read into R, calculate delta and also total observations...

---

# Heterozygosity
.tiny[



```r
# Load heterozygosity 
atlantic_raw_het_freq &lt;- read_tsv(atlantic_het_path)
baltic_raw_het_freq &lt;- read_tsv(baltic_het_path)

# Separate `OBS(HOM1/HET/HOM2)` column, calculate frequency and select important
atlantic_het_freq &lt;- atlantic_raw_het_freq %&gt;% 
  separate(`OBS(HOM1/HET/HOM2)`, c("OBS_HOM1", "OBS_HET", "OBS_HOM2"), convert = T) %&gt;%
* mutate(TOT_OBS = OBS_HOM1 + OBS_HET + OBS_HOM2,
*        FREQ = OBS_HET / TOT_OBS,
         ID = "Atlantic Obs Het Freq") %&gt;%
  select(POS, FREQ, ID, TOT_OBS)

baltic_het_freq &lt;- baltic_raw_het_freq %&gt;% 
  separate(`OBS(HOM1/HET/HOM2)`, c("OBS_HOM1", "OBS_HET", "OBS_HOM2"), convert = T) %&gt;%
  mutate(TOT_OBS = OBS_HOM1 + OBS_HET + OBS_HOM2, 
         FREQ = OBS_HET / TOT_OBS,
         ID = "Baltic Obs Het Freq") %&gt;%
  select(POS, FREQ, ID, TOT_OBS)

# Initialize delta df with POS
delta_het_freq &lt;- atlantic_het_freq %&gt;% select(POS)

# Calculate delta 
delta_het_freq &lt;- delta_het_freq %&gt;% 
  mutate(FREQ = abs(atlantic_het_freq$FREQ - baltic_het_freq$FREQ), 
         ID = "Delta Obs Het Freq")

atlantic_baltic_het_freq &lt;- rbind(atlantic_het_freq, baltic_het_freq)
```
]

---

# Missing observations 


```r
atlantic_baltic_het_freq %&gt;%
  group_by(gr = cut(POS, break_size(POS, 1e5)), ID) %&gt;%
  summarize(`Position (Mb)` = mean(POS/1e6), `Mean missingness` = mean(1-TOT_OBS/max(TOT_OBS))) %&gt;%
  ggplot(aes(`Position (Mb)`,`Mean missingness`, col = ID)) + 
  geom_line()
```

![](herring_project_presentation_files/figure-html/missingness-plot-1.png)&lt;!-- --&gt;

---

# Filter the data


```r
# Choose a filter threshold, % of max
filter_threshold = 1

# Keep positions 
positions_to_keep &lt;- inner_join(atlantic_het_freq, baltic_het_freq, by = "POS") %&gt;%
  filter(
*   TOT_OBS.x &gt;= filter_threshold * max(TOT_OBS.x) &amp;
*   TOT_OBS.y &gt;= filter_threshold * max(TOT_OBS.y)
    ) %&gt;% 
  select(POS)
```



---

# After filtering


```r
combined_allele_freq %&gt;%
* filter(POS %in% positions_to_keep$POS) %&gt;%
  group_by(gr = cut(POS, break_size(POS, window_size)), ID) %&gt;% 
  summarize(`Position (Mb)` = mean(POS/1e6), `Mean Frequency` = mean(FREQ)) %&gt;%
  ggplot(aes(`Position (Mb)`,`Mean Frequency`, col = ID)) + 
  geom_jitter(width = 0, height = 0.03, size = .1, alpha = .3) + 
  facet_wrap(~ID, ncol = 3)
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-12-1.png)&lt;!-- --&gt;

???

Mention here the delta-freq peak

---

# Heterozygosity after filtering


```r
*combined_het_freq %&gt;%
  filter(POS %in% positions_to_keep$POS) %&gt;% 
  group_by(gr = cut(POS, break_size(POS, window_size)), ID) %&gt;% 
  summarize(`Position (Mb)` = mean(POS/1e6), `Mean Frequency` = mean(FREQ)) %&gt;%
  ggplot(aes(`Position (Mb)`,`Mean Frequency`, col = ID)) + 
  geom_jitter(width = 0, height = 0.03, size = .1, alpha = .3) + 
  facet_wrap(~ID, ncol = 3)
```

![](herring_project_presentation_files/figure-html/unnamed-chunk-13-1.png)&lt;!-- --&gt;

---

# Sweep region 

![](herring_project_presentation_files/figure-html/unnamed-chunk-14-1.png)&lt;!-- --&gt;

---

# Annotation


```bash
# To get the annotations from the VCF-file
bcftools query -f "%POS\t%ANN\n" chr4.vcf &gt; output.txt
```




```r
# Reads the annotation file really really fast 
library(vroom) 

annotation &lt;- vroom(annotations_path, delim = "\t", col_names = c("POS", "ANN"))

# Annotate delta frequencies
annotated_positions &lt;- delta_allele_freq %&gt;% 
  # In the approximate region
* filter(between(POS, 1e7, 1.3e7)) %&gt;%
  # With a high delta freq
* filter(FREQ &gt; 0.5) %&gt;%
  # Keep only SNPs with good data
* filter(POS %in% positions_to_keep$POS) %&gt;%
  # Run the annotation function
* annotate_positions(annotation)
```

???

I could take the annotations parts maybe? (and more if needed)

---

# annotate_positions()

.tiny[

```r
annotate_positions &lt;- function(input_df, annotation_df) {
  
  # Filter positions which should be annotated 
  filtered_annotations &lt;- annotation_df %&gt;%   
    filter(POS %in% input_df$POS)
  
  # This info can be found in the VCF file 
  snpEff_column_names &lt;- c("Allele","Annotation","Annotation_Impact","Gene_Name","Gene_ID",
                           "Feature_Type","Feature_ID","Transcript_BioType","Rank",
                           "HGVS.c","HGVS.p","cDNA.pos / cDNA.length","CDS.pos / CDS.length",
                           "AA.pos / AA.length","Distance","ERRORS / WARNINGS / INFO")   
  
  untangled_annotations &lt;- filtered_annotations %&gt;%
    mutate(ANN = str_split(ANN, ",")) %&gt;%         # Split the ANN column into a list of annotations
    unnest(ANN) %&gt;%                               # create rows for each position - annotation pair
    group_by(POS) %&gt;%                             # group by position
    mutate(Annotation_ID = seq_along(POS), .after = POS) %&gt;% # assign a number to each annotation
    ungroup() %&gt;%                                 # ungroup the values again 
    separate(ANN, into = snpEff_column_names,     # separate the ANN column
             sep = "\\|")                         # into columns based on SnpEff information
  
  # Add annotations to input dataframe 
  output_df &lt;- inner_join(input_df, untangled_annotations, by = "POS") 
  
  return(output_df)
}
```
]

???

I could take the annotations parts maybe? (and more if needed)

---

# Annotations

Before
.nowrap.tiny[

```r
delta_allele_freq %&gt;% print(n = 4)
```

```
## # A tibble: 1,158,327 x 3
##     POS   FREQ ID               
##   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;            
## 1   369 0.0227 Delta Allele Freq
## 2   404 0      Delta Allele Freq
## 3   421 0.0357 Delta Allele Freq
## 4   427 0.0185 Delta Allele Freq
## # … with 1,158,323 more rows
```
]
After
.nowrap.tiny[

```r
annotated_positions %&gt;% print(n = 4)
```

```
## # A tibble: 258 x 20
##       POS  FREQ ID    Annotation_ID Allele Annotation Annotation_Impa… Gene_Name
##     &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;         &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;    
## 1  1.04e7 0.564 Delt…             1 G      intron_va… MODIFIER         ENSCHAG0…
## 2  1.04e7 0.564 Delt…             2 G      intron_va… MODIFIER         ENSCHAG0…
## 3  1.04e7 0.564 Delt…             3 G      intron_va… MODIFIER         ENSCHAG0…
## 4  1.04e7 0.564 Delt…             4 G      intron_va… MODIFIER         ENSCHAG0…
## # … with 254 more rows, and 12 more variables: Gene_ID &lt;chr&gt;,
## #   Feature_Type &lt;chr&gt;, Feature_ID &lt;chr&gt;, Transcript_BioType &lt;chr&gt;, Rank &lt;chr&gt;,
## #   HGVS.c &lt;chr&gt;, HGVS.p &lt;chr&gt;, cDNA.pos / cDNA.length &lt;chr&gt;,
## #   CDS.pos / CDS.length &lt;chr&gt;, AA.pos / AA.length &lt;chr&gt;, Distance &lt;chr&gt;,
## #   ERRORS / WARNINGS / INFO &lt;chr&gt;
```
]

---

# Filter the output

.tiny[

```r
# Can then filter what we want
annotated_positions %&gt;% group_by(Gene_ID) %&gt;%
  filter(Feature_Type == "transcript") %&gt;%
  select(Gene_ID, POS, FREQ) %&gt;% unique() %&gt;%
  summarize(mean.POS = as.integer(mean(POS)), mean.FREQ = mean(FREQ))
```

```
## # A tibble: 15 x 3
##    Gene_ID            mean.POS mean.FREQ
##    &lt;chr&gt;                 &lt;int&gt;     &lt;dbl&gt;
##  1 ENSCHAG00000006163 10374208     0.564
##  2 ENSCHAG00000025774 11175669     0.634
##  3 ENSCHAG00000025873 11219606     0.608
##  4 ENSCHAG00000025879 11292290     0.629
##  5 ENSCHAG00000025886 11335365     0.643
##  6 ENSCHAG00000025895 11357239     0.620
##  7 ENSCHAG00000026024 11380240     0.575
##  8 ENSCHAG00000026767 11411895     0.602
##  9 ENSCHAG00000026775 11439566     0.571
## 10 ENSCHAG00000026777 11458459     0.571
## 11 ENSCHAG00000026796 11484890     0.546
## 12 ENSCHAG00000026905 11547797     0.567
## 13 ENSCHAG00000026909 11581898     0.571
## 14 ENSCHAG00000026911 11595762     0.550
## 15 ENSCHAG00000026916 11637637     0.536
```
]

???

I could take the annotations parts maybe? (and more if needed)

---



# biomaRt

Used biomaRt to get gene names for each Gene ID

.tiny[

```r
# A bug in biomaRt overwrites dplyr::select, workaround is to unload dplyr, then load it _after_ biomaRt
#detach("package:dplyr")
#library(biomaRt)
#library(dplyr)

# connect to the herring database
ensembl = useMart("ensembl",dataset="charengus_gene_ensembl")

annotated_gene_names &lt;- annotated_positions %&gt;%            # Take our annotated positions
  group_by(Gene_ID) %&gt;%                        # Group by Gene_ID
  filter(Feature_Type == "transcript" &amp;        # Filter for transcripts
        Annotation != "intron_variant") %&gt;%    # Remove intron_variants (remove "double-ids")
  select(POS, FREQ, Gene_ID, ID) %&gt;%           # Select columns to keep 
  unique() %&gt;%                                 # Don't let multiple transcripts affect the means
  summarize(mean.pos = mean(POS),              # Summarize mean position and frequency
            mean.freq = mean(FREQ), 
            Gene_ID, ID) %&gt;% 
  unique() %&gt;%                                 # keep only unique rows
  mutate(getBM(attributes = c('ensembl_gene_id', 'external_gene_name'), # New columns from ensambl
               filters = 'ensembl_gene_id',    # where the ensembl_gene_id
               values = Gene_ID,               # equals the Gene_ID
               mart = ensembl,                 # chose mart/database
               uniqueRows = F)) %&gt;%            # give us everything
  mutate(external_gene_name = ifelse(external_gene_name == "", NA, external_gene_name)) %&gt;% # change "" to NA
  #na.omit() %&gt;%                                # keep only genes with zfin_id's
  select(-ensembl_gene_id)                     # remove duplicate column
```
]

???

I could take the annotations parts maybe? (and more if needed)

---

# biomaRt

.tiny[

```r
annotated_gene_names %&gt;% 
  #filter(external_gene_name != "NA") %&gt;% 
  arrange(-mean.freq)
```

```
## # A tibble: 13 x 4
## # Groups:   Gene_ID [13]
##    Gene_ID             mean.pos mean.freq ID               
##    &lt;chr&gt;                  &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;            
##  1 ENSCHAG00000025879 11326441      0.643 Delta Allele Freq
##  2 ENSCHAG00000025886 11329910.     0.643 Delta Allele Freq
##  3 ENSCHAG00000025895 11352886.     0.636 Delta Allele Freq
##  4 ENSCHAG00000026024 11384018.     0.629 Delta Allele Freq
##  5 ENSCHAG00000025774 11179139      0.614 Delta Allele Freq
##  6 ENSCHAG00000025873 11219607.     0.608 Delta Allele Freq
##  7 ENSCHAG00000026767 11416264.     0.605 Delta Allele Freq
##  8 ENSCHAG00000026775 11439566      0.571 Delta Allele Freq
##  9 ENSCHAG00000026777 11458460.     0.571 Delta Allele Freq
## 10 ENSCHAG00000026909 11581898      0.571 Delta Allele Freq
## 11 ENSCHAG00000026911 11594816      0.557 Delta Allele Freq
## 12 ENSCHAG00000026796 11488153      0.552 Delta Allele Freq
## 13 ENSCHAG00000026916 11637637      0.536 Delta Allele Freq
```
]

---

# biomaRt

.center.top[
![ensembl](ensembl.png)]

???

I could take the annotations parts maybe? (and more if needed)

---



# Summary plot 

![](herring_project_presentation_files/figure-html/plot-delta-gene-names-1.png)&lt;!-- --&gt;



???

This as well (if needed)

---

# Conclusions

- Results show that only one selective sweep occurred in chromosome 4 of the baltic herring genome

- This occurred between approximately 11-12 Mb

- The genes involved in the sweep were magi1a, rho, qars1, espn, ENSCHAG00000025879, ENSCHAG00000025886 and ENSCHAG00000025895 and became fixed in the new population.

- ENSCHAG00000025879, ENSCHAG00000025886 and ENSCHAG00000025895 have poorly defined functions

- magi1a, qars1, espn are involved in signal transduction, glutamine-tRNA ligase activity and actin filament binding activity respectively. Hard to understand how their mutations will affect the population without further investigation

---

# Conclusions

- rho creates G-protein-coupled photoreceptors used as light receptors

- Hill et al. showed replacement of phenylalanine by tyrosine at position 261 in rhodopsin. This allows better absorption of red wavelengths.

- The Baltic Sea has a red shifted light environment compared to the Atlantic Ocean which is blue shifted. The baltic herring fixed a mutation allowing them to see better in their new environment. 

---

class: center, top
background-image: url("https://upload.wikimedia.org/wikipedia/commons/c/c5/Clupea_harengus_Gervais.flipped.jpg")
background-size: contain

# Thanks for listening! 
.center[(_and for all the fish_)]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
